File: Sources/Applicazione/main/java/it/kaizenteam/app/model/NorrisChart/BarChartInPlaceUpdater.java
/*
 * Name: BarChartInPlaceUpdater
 * Package: it.kaizenteam.app.model.NorrisChart
 * Location: Sources/Applicazione/main/java/it/kaizenteam/app/model/NorrisChart
 * Date: 24/05/2015
 * Version: 0.01
 *
 * History:
 * =================================================================
 * Version	Date	Programmer	Changes
 *  =================================================================
 * v0.01	2015-05-24	Alessandro Moretto	Creazione file
 * =================================================================
 *
 */


package it.kaizenteam.app.model.NorrisChart;

/**
 * This class is responsible for defining the in place update method for a bar chart. It can access to the BarChartImpl private data fields because it is an inner class of BarChartImpl. In particular, it can access to DataObject contained in  BarChartImpl and change its values.
 */
public class BarChartInPlaceUpdater implements Updater {
    /*
     * The static attribute is the unique instance of that class.
     */
    private static BarChartInPlaceUpdater instance;

    /**
     * This method is responsible for returning the unique instance of this class and creating it if it not exists.
     * @return the unique instance of the class
     */
    public static Updater getInstance(){
        return instance;
    }

    /**
     * Constructor
     */
    private BarChartInPlaceUpdater(){}

    /**
     * This method has the task of updating the chart as a parameter (chart) by using the update package (UpdateData).
     * @param chart
     * @param updateData update package
     */
    @Override
    public void update(ChartImpl chart, ChartUpdate updateData) {
//TODO
    }
}

File: Sources/Chuck/main/model/norrischart/BarChartInPlaceUpdater.js
/*
 * Name: {BarChartInPlaceUpdater.js}
 * Module: {}
 * Location: {Norris/Main/DataModel}
 * Date: {2015-05-15}
 * Version: {v0.1}
 *
 * History:
 *
 * ================================================================================
 * Version Date Programmer Changes
 * ================================================================================
 * 0.01 2015-05-15 Chiara Bigarella {creation}
 * ================================================================================
 */

angular.module('norris-chartupdater', [])

.factory('BarChartInPlaceUpdater',function () {

    function BarChartInPlaceUpdater() {
        if(!(this instanceof BarChartInPlaceUpdater)) return new BarChartInPlaceUpdater();
    }

    BarChartInPlaceUpdater.prototype.instance=new BarChartInPlaceUpdater(); // static

    BarChartInPlaceUpdater.getInstance = function() { // static
        return BarChartInPlaceUpdater.prototype.instance;
    };

    /**
     * Updates a bar chart with in place method. The bar chart data should not be empty.
     * @param {ChartImpl} chart - the bar chart to update;
     * @param updateData - the updating.
     */
    BarChartInPlaceUpdater.prototype.update = function (chart, updateData) {
        var isEmpty=function(obj) {
        for(var prop in obj) {
            if(obj.hasOwnProperty(prop))
                return false;
        }
        return true;
        };

        if (!isEmpty(updateData)) {
            var data=chart.getData();
            if (!isEmpty(data)) {
                for(var i=0; i<updateData.length; i++) {
                    var x=updateData[i].position.x;
                    var y=updateData[i].position.y;
                    data.datasets[x].values[y]=updateData[i].value;
                }
                chart.setData(data);
            }
            else {
                console.log("ERROR: the chart has no data to update.");
                throw ("emptyChart");
            }
        };
    };

    return BarChartInPlaceUpdater;

});